### PR TITLE
Creates the base files for generating go protobufs

### LINKED ISSUE #2: Create minimal code generation tooling
https://github.com/open-telemetry/opentelemetry-proto-go/pull/1

### PR DESCRIPTION
**Pull request by MadVikingGod**
# Summary
This PR aims to start the conversation around putting the generated protobuf files into a consumable module.  

There are several places where these files are generated but are done internally so that other projects can't consume them.  This takes the [generation code from opentelemetry-go](https://github.com/open-telemetry/opentelemetry-go/blob/master/Makefile.protos) as a base to generate the protobufs that will eventually be consumed.

# Things not addressed

- Vanity URLs.  I have assumed that the use of `go.opentelemetry.io/proto`.  This is easily changed.
- The actually generated files.  These are large and can be added after reviewing the technical parts of generation are complete.
- Tests, e.g., that the source and the generated files match.  Automation of this kind can be added after this proof of concept
- Notification/Automation, e.g., some signal that this repository is out of date and should be updated.
- Updating the proto generation tool.  Again, this can be addressed after this is shown to be useful.
 

Resolves #2

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+PROTOC_ZIP_URL := https://github.com/protocolbuffers/protobuf/releases/download/v$(PROTOC_VERSION)/protoc-$(PROTOC_VERSION)-$(PROTOC_OS)-$(PROTOC_ARCH).zip
+
+$(TOOLS_DIR)/PROTOC_$(PROTOC_VERSION):
+	@rm -f "$(TOOLS_DIR)"/PROTOC_* && \
+	touch "$@"
+
+# Depend on a versioned file (like PROTOC_3.14.0), so when version
+# gets bumped, we will depend on a nonexistent file and thus download
+# a newer version.
+$(TOOLS_DIR)/protoc/bin/protoc: $(TOOLS_DIR)/PROTOC_$(PROTOC_VERSION)
```
Comment (bogdandrutu) [inline Makefile:73]: please use the docker container for this https://github.com/open-telemetry/opentelemetry-proto/blob/master/makefile

Comment (MadVikingGod) [reply Makefile]: Good call.  That simplified things quite a bit.

**Comment by punya** [comment]
Did we consider the approach of generating language-specific packages from a Github action in opentelemetry-proto, rather than creating a separate repo that includes the protos as a submodule, needs to be updated, etc.? In the case of Go, the generated artifact would need to live in its own Github repo of course, but that repo would be *purely* for automated updates.

**Comment by MadVikingGod** [comment]
My understanding is that the maintainers of opentelemetry-proto do not want to support N-language generation in that repository. It is up to the individual language SIGs to support this kind of effort.

**Comment by MrAlias** [comment]
> Did we consider the approach of generating language-specific packages from a Github action in opentelemetry-proto, rather than creating a separate repo that includes the protos as a submodule, needs to be updated, etc.? In the case of Go, the generated artifact would need to live in its own Github repo of course, but that repo would be _purely_ for automated updates.

@punya yes we have discussed this. Please see https://github.com/open-telemetry/opentelemetry-proto/issues/79 for context.

File: go.mod [hunk 2]
```diff
@@ -0,0 +1,3 @@
+module go.opentelemetry.io/proto
```
Comment (MrAlias) [inline go.mod:1]: I think `go.opentelemetry.io/proto` could work, but `go.opentelemetry.io/otlp` might be more descriptive of what the contents is.

**Review by MrAlias — APPROVED** [review]

File: go.mod (hunk 2 above)
Comment (MadVikingGod) [reply go.mod:1]: So I'm happy to change that, but I would also recommend that we would include `proto` somewhere in the path.  Just to keep the expectation that you won't find any kind of exporter logic or other such affordances in these packages.  They are just the protobuf serde code, and the client and server stubs.

Comment (MrAlias) [reply go.mod:1]: From the SIG meeting, the proposed module name is `go.opentelemetry.io/proto/otlp`

File: README.md [hunk 3]
```diff
@@ -1,2 +1,18 @@
 # opentelemetry-proto-go
 Generated code for OpenTelemetry protobuf data model
+
+# Usage
+You can import the generated code directly in your project
+```
+import (
+	tracepb "go.opentelemetry.io/proto/trace/v1"
```
Comment (MrAlias) [inline README.md:8]:
```suggestion
	tracepb "go.opentelemetry.io/proto/otlp/trace/v1"
```

File: README.md [hunk 4]
```diff
@@ -1,2 +1,18 @@
...
+)
+```
+
+# Generating new versions
+When [opentelemeetry-proto](https://github.com/open-telemetry/opentelemetry-proto) release a new version we can update submodule, and then regenerate the protobufs.
```
Comment (MrAlias) [inline README.md:13]:
```suggestion
When [opentelemetry-proto](https://github.com/open-telemetry/opentelemetry-proto) release a new version we can update submodule, and then regenerate the protobufs.
```

**Review by MrAlias — COMMENTED** [review]
Can we checkout the proto submodule into an internal package so external dependencies on that code are made here?

**Comment by MadVikingGod** [comment]
Fixed the readme.  What is the ask with the proto files?

Do you want the submodule which holds the `.proto` files in its own module?
Is this about the code that exists in `gen/`?  I wasn't expecting to check that in, but it could be generated in an internal directory or module just to make sure its dependencies aren't accidentally checked into the root `go.mod`.

**Comment by MrAlias** [comment]
> What is the ask with the proto files?
> 
> Do you want the submodule which holds the `.proto` files in its own module?
> Is this about the code that exists in `gen/`? I wasn't expecting to check that in, but it could be generated in an internal directory or module just to make sure its dependencies aren't accidentally checked into the root `go.mod`.

Looking a bit more into the ask I'm not sure it is valid, but here's a description of what I was thinking. By creating the opentelemetry-proto submodule at a top-level users will be able to `import go.opentelemetry.io/proto/opentelemetry-proto/...`.

I took another look at the proto repo and it doesn't look like there is any Go code in the repo anymore so this is probably not a concern. However, it might make things a bit cleaner if instead of instantiating the opentelemetry-proto submodule at a path of `./opentelemetry-proto` we create it at `./internal/opentelemetry-proto` so there isn't any confusion in the future if that repo starts to contain Go code (probably overkill, so feel free to disregard).

**Comment by MadVikingGod** [comment]
Just to add to that, even if there were go code its mod file wouldn't match that directory, and you would get an import mismatch.  

Playing around I moved `opentelemetry-proto` and `gen` into a subfolder, I called `proto` and added a mod with a non-canonical name so it couldn't be imported.  So we then had `proto\opentelemetry-proto` and `proto\gen`.  This made the structure a bit more obvious, but I also don't like the name `proto` for this purpose.

I don't think it matters for this, as this is all non-imported or persisted code.  I also think that we would be better served by a CI check that the root go.mod doesn't change.  I.E. there shouldn't be anything imported by the root level, as there is no code.

**Comment by MrAlias** [comment]
All good points and ideas. Sounds like we can disregard my comment about moving the path. I think we can include that check you described at a later time as well so as to not have the scope of this PR creep too much.

**Comment by MrAlias** [comment]
@bogdandrutu any other thoughts before we merge this?

**Pull request merged**